README
Just the start.


NEWS
Version 1.0.0
-----------------------------------------------------------------------------------------------------
First functions.

test/oldRoxDoc.r
#' facet_wrap_helper
#' 
#' Call facet_wrap
#' 
#' This function merely serves to call \code{\link{facet_wrap}}.  It is needed because the program doesn't know if \code{\link{facet_wrap}} or \code{\link{facet_grid}} is being called and hence doesn't know what arguments to pass, and those functions are not equiped with \dots.
#' 
#' @author Jared P. Lander
#' @aliases facet_wrap_helper facet_grid_helper none_helper
#' @seealso facet_wrap facet_grid
#' @import ggplot2
#' @param formula \code{\link{formula}} for use in \code{link{facet_wrap}}
#' @param nrow Number of rows.
#' @param ncol Number of columns.
#' @param scales Should scales be fixed ("fixed", the default), free ("free"), or free in one dimension ("free_x", "free_y").
#' @param \dots Catch all.
#' @return The result of \code{link{facet_wrap}}.
#'



#' facet_grid_helper
#' 
#' Call facet_grid
#' 
#' This function merely serves to call \code{\link{facet_grid}}.  It is needed because the program doesn't know if \code{\link{facet_wrap}} or \code{\link{facet_grid}} is being called and hence doesn't know what arguments to pass, and those functions are not equiped with \dots.
#' 
#' @author Jared P. Lander
#' @aliases facet_wrap_helper facet_grid_helper none_helper
#' @seealso facet_wrap facet_grid
#' @import ggplot2
#' @param formula \code{\link{formula}} for use in \code{link{facet_grid}}
#' @param scales Should scales be fixed ("fixed", the default), free ("free"), or free in one dimension ("free_x", "free_y").
#' @param \dots Catch all.
#' @return The result of \code{link{facet_grid}}.



#' none_helper
#' 
#' Call none
#' 
#' This function merely serves to call \code{]link{none}}.  It is needed because the program doesn't know if \code{]link{facet_wrap}} or \code{\link{facet_grid}} is being called and hence doesn't know what arguments to pass, and those functions are not equiped with \dots.
#' 
#' @author Jared P. Lander
#' @aliases facet_wrap_helper facet_grid_helper none_helper
#' @seealso facet_wrap facet_grid none
#' @import ggplot2
#' @param \dots Catch all.
#' @return \code{\link{NULL}}.
#'

r/shape.r
#' prepare.shape.fortify
#' 
#' Prepare Shape File for Plotting
#' 
#' Takes in a shape object, fortifies it and joins in it's accompanying data.
#' 
#' @param shape A shape object as read using maptools.
#' @return A \code{\link{data.frame}} suitable for use in \code{ggplot}.
#' @author Jared P. Lander
#' @aliases prepare.shape.fortify
#' @export prepare.shape.fortify
#' @import ggplot2 plyr maptools
#' @seealso prepare.shape
#' @examples \dontrun{
#' nyc <- readShapeSpatial(fn="nyct2010.shp")
#' nyc.df <- prepare.shape.fortify(nyc)
#' }
#' 
prepare.shape.fortify <- function(shape)
{
    # needed for fortify to work
    suppressMessages(gpclibPermit())
    # put rownames as id in the data
    shape@data$id <- rownames(shape@data)
    # fortify the shape file
    shape.points <- fortify(shape, region="id")
    # join in info from data
    shape.df <- join(shape.points, shape@data, by="id")
    #housekeeping
    rm(shape, shape.points)
    
    return(shape.df)
}


#' prepare.shape
#' 
#' Prepare Shape File for Plotting
#' 
#' Reads in a shape file and then fortifies it.
#' @author Jared P. Lander
#' @seealso prepare.shape.fortify
#' @export prepare.shape
#' @aliases prepare.shape
#' @param file Name of shape file.
#' @return A \code{\link{data.frame}} suitable for use in \code{ggplot}.
#' @examples \dontrun{
#' nyc.df <- prepare.shape("nyct2010.shp")
#' }
#' 
prepare.shape <- function(file)
{
    # read shape file
    shape <- readShapeSpatial(fn=file)
    # frotify and join the file
    prepare.shape.fortify(shape)
}

r/mapping.r
# mapping function
#' map.plot
#' 
#' Plot Maps
#' 
#' Takes a \code{\link{data.frame}} containing shapefile data and plots it along with some measure.
#' 
#' Will probably need to make changes for when plotting continuous data.
#' 
#' @author Jared P. Lander
#' @aliases map.plot
#' @export map.plot
#' @import ggplot2 scales
#' @param data \code{\link{data.frame}} resulting from using \code{\link{fortify}} on a shapefile object.
#' @param variable Character indicating which column should be plotted.
#' @param fill.color.high Color to use for top of gradient.
#' @param space Color space to use for gradient.
#' @param path.color Color to use for shapefile lines.
#' @param title Title of plot.
#' @param xlab X-axis label.
#' @param ylav Y-axis label.
#' @param barheight Height of legend bar.
#' @param formatter How the legend numbers should be formatted.  Takes a function.
#' @param legend.position Position of legend.
#' @return A ggplot object.
#' @examples
#' 
#' \dontrun{
#' nyc.income <- join(nyc.df, income2010, by="Key")
#' mana <- nyc.income[nyc.income$County == 61, ]
#' map.plot(mana, "Median.Income", formatter=multiple_format(extra=dollar, multiple="K"))
#' }
#' 
map.plot <- function(data, variable, fill.color.high=muted("green"), space="Lab", path.color="white", title=NULL, 
                     xlab=NULL, ylab=NULL, barheight=15, formatter=multiple_format(multiple="K", extra=comma),
                     legend.position=c("right", "bottom", "left", "top", "none"))
{
    # generate a bunch of prebuilt map options for easier plotting
    mapOpts <- map.options()
    
    # get legend.position
    legend.position <- match.arg(legend.position)
    
    # start plotting
    p <- ggplot(data) + 
        # lat/long and group
        aes(x=long, y=lat, group=group) + 
        # coloring based on the variable
        aes_string(fill=variable) +  geom_polygon() + 
        geom_path(color=path.color) + coord_equal() + 
        # mapping options
        mapOpts + 
        # x/y labels
        labs(x=xlab, y=ylab) + 
        # make the legend be titleless and have a long, tal color bar
        guides(fill=guide_colorbar(title=NULL, ticks=FALSE, barheight=barheight)) + 
        # make the color scale a gradient and use the chosen formatter
        scale_fill_gradient2(labels=formatter, space=space, high=fill.color.high) + 
        # title of plot
        opts(title=title)
    
    return(p)
}

#' map.options
#' 
#' Common mapping options
#' 
#' This builds a ggplot opts list that is commonly used for plotting.  Currently it only gives the default, perhaps I'll make it so users can pick and choose.
#' 
#' @author Jared P. Lander
#' @aliases map.options
#' @export map.options
#' @import ggplot2
#' @return A \code{\link{list}} of ggplot items.
map.options <- function()
{
    list(opts(panel.grid.major=theme_blank(), panel.grid.minor=theme_blank(), axis.text.x=theme_blank(), axis.text.y=theme_blank(), axis.ticks=theme_blank(), panel.background=theme_blank()))
}

r/formatters.r
## formatting functions

#' multiple
#' 
#' Order of Magnitude Formatter
#' 
#' This divides the number by the apropriate amount and adds on the corresponding symbol at the end of the number.
#' 
#' @author Jared P. Lander
#' @aliases multiple
#' @export multiple
#' @import scales
#' @param x Vector of numbers to be formatted.
#' @param The multiple to display numbers in.  This symbol will be added to the end of the numbers.
#' @param extra Function for perform any further formatting.
#' @param digits Number of decimal places for rounding.
#' @return Character vector of formatted numbers.
#' @examples
#' vect <- c(1000, 1500, 23450, 21784, 875003780)
#' multiple(vect)
#' multiple(vect, extra=dollar)
#' multiple(vect, extra=identity)
#' 
multiple <- function(x, multiple=c("K", "M", "B", "T", "H", "k", "m", "b", "t", "h"), extra=comma, digits=0)
{
    # get the multiple
    multiple=match.arg(multiple)
    
    # set up a vector for dividing
    dividers <- c("K"=1000, "M"=1000000, "B"=1000000000, "T"=1000000000000, "H"=100)
    
    # get what we're dividing by
    divider <- dividers[toupper(multiple)]
    
    x <- round(x / divider, digits=digits)
    
    x <- do.call(extra, args=list(x))
    sprintf("%s%s", x, multiple)
}


#' multiple_format
#' 
#' Multiple Style Formatting
#' 
#' Since ggplot requires a function for formatting this allows the user to specify the function's arguments, which will return a function that can be used by ggplot.
#' 
#' @param \dots Arguments to be passed onto \code{\link{multiple}}
#' @return The function \code{\link{multiple}}.
#' @author Jared P. Lander
#' @export multiple_format
#' @aliases multiple_format
#' @examples
#' 
#' vect <- c(1000, 1500, 23450, 21784, 875003780)
#' multiple_format()(vect)
#' multiple_format(extra=dollar)(vect)
#' multiple_format(extra=identity)(vect)
multiple_format <- function(...)
{
    function(x) multiple(x, ...)
}



#' multiple.dollar
#' 
#' Order of Magnitude Formatter
#' 
#' Simply a wrapper for multiple that prespecifies the extra dollar.
#' 
#' @author Jared P. Lander
#' @aliases multiple.dollar
#' @export multiple.dollar
#' @import scales
#' @param x Vector of numbers to be formatted.
#' @param The multiple to display numbers in.  This symbol will be added to the end of the numbers.
#' @param digits Number of decimal places for rounding.
#' @return Character vector of dollar formatted numbers.
#' @examples
#' vect <- c(1000, 1500, 23450, 21784, 875003780)
#' multiple.dollar(vect)
#' multiple.dollar(vect, multiple="k")
#' multiple.dollar(vect, multiple="h")
#'
multiple.dollar <- function(x, multiple=c("K", "M", "B", "T", "k", "m", "b", "t"))
{
    multiple(x=x, multiple=multiple, extra=dollar)
}



#' multiple.comma
#' 
#' Order of Magnitude Formatter
#' 
#' Simply a wrapper for multiple that prespecifies the extra comma.
#' 
#' @author Jared P. Lander
#' @aliases multiple.comma
#' @export multiple.comma
#' @import scales
#' @param x Vector of numbers to be formatted.
#' @param The multiple to display numbers in.  This symbol will be added to the end of the numbers.
#' @param digits Number of decimal places for rounding.
#' @return Character vector of comma formatted numbers.
#' @examples
#' vect <- c(1000, 1500, 23450, 21784, 875003780)
#' multiple.comma(vect)
#' multiple.comma(vect, multiple="k")
#' multiple.comma(vect, multiple="h")
#'
multiple.comma <- function(x, multiple=c("K", "M", "B", "T", "k", "m", "b", "t"))
{
    multiple(x=x, multiple=multiple, extra=comma)
}



#' multiple.comma
#' 
#' Order of Magnitude Formatter
#' 
#' Simply a wrapper for multiple that prespecifies the extra identity.
#' 
#' @author Jared P. Lander
#' @aliases multiple.identity
#' @export multiple.identity
#' @import scales
#' @param x Vector of numbers to be formatted.
#' @param The multiple to display numbers in.  This symbol will be added to the end of the numbers.
#' @param digits Number of decimal places for rounding.
#' @return Character vector of formatted numbers.
#' @examples
#' vect <- c(1000, 1500, 23450, 21784, 875003780)
#' multiple.identity(vect)
#' multiple.identity(vect, multiple="k")
#' multiple.identity(vect, multiple="h")
#'
multiple.identity <- function(x, multiple=c("K", "M", "B", "T", "k", "m", "b", "t"))
{
    multiple(x=x, multiple=multiple, extra=identity)
}

r/plotting.r
# mapping function
#' map.plot
#' 
#' Plot Maps
#' 
#' Takes a \code{\link{data.frame}} containing shapefile data and plots it along with some measure.
#' 
#' Will probably need to make changes for when plotting continuous data.
#' 
#' @author Jared P. Lander
#' @aliases map.plot
#' @export map.plot
#' @import ggplot2 useful
#' @param data \code{\link{data.frame}} resulting from using \code{\link{fortify}} on a shapefile object.
#' @param variable Character indicating which column should be plotted.
#' @param longitude Character indicating which column stores longitude values.
#' @param latitude Character indicating which column stores latitude values.
#' @param fill.color.high Color to use for top of gradient.
#' @param space Color space to use for gradient.
#' @param path.color Color to use for shapefile lines.
#' @param title Title of plot.
#' @param title.size Size of \code{Title} font.
#' @param title.hjust Horizontal adjustment of \code{Title} position.
#' @param xlab X-axis label.
#' @param ylab Y-axis label.
#' @param barheight Height of legend bar.
#' @param formatter How the legend numbers should be formatted.  Takes a function.
#' @param legend.position Position of legend.
#' @param lhs Left side of formula to use for facetting.
#' @param rhs Right side of formula to use for facetting.
#' @param facet The type of facetting, if any, to use.
#' @param wrap.nrow The number of rows to to use when using facet_wrap.
#' @param wrap.ncol The number of columns to to use when using facet_wrap.
#' @param scales The scales to be employed when faceting.
#' @return A ggplot object.
#' @examples
#' 
#' \dontrun{
#' nyc.income <- join(nyc.df, income2010, by="Key")
#' mana <- nyc.income[nyc.income$County == 61, ]
#' map.plot(mana, "Median.Income", formatter=multiple_format(extra=dollar, multiple="K"))
#' }
#' 
map.plot <- function(data, variable, longitude="long", latitude="lat", fill.color.high=muted("green"), space="Lab", 
                     path.color="white", 
                     title=NULL, title.size=15, title.hjust=.5,
                     xlab=NULL, ylab=NULL, barheight=15, formatter=percent,
                     legend.position=c("right", "bottom", "left", "top", "none"), 
                     lhs=NULL, rhs=NULL, facet=c("none", "facet_wrap", "facet_grid"), wrap.nrow=NULL, wrap.ncol=NULL, 
                     scales=c("fixed", "free", "free_y", "free_x")
                     )
{
    # old formatter multiple_format(multiple="K", extra=comma)
    # generate a bunch of prebuilt map options for easier plotting
    mapOpts <- map.options()
    
    # get legend.position
    legend.position <- match.arg(legend.position)
    
    # get arguments about faceting
    facet <- match.arg(facet)
    scales <- match.arg(scales)
    
    # build the string to call the helper functions
    facet <- sprintf("%s_helper", facet)
    
    # build facets, could be NULL or facet_wrap or facet_grid
    facet <- do.call(facet, args=list(formula=build.formula(lhs=lhs, rhs=rhs), nrow=wrap.nrow, ncol=wrap.ncol, scales=scales))
    
    # start plotting
    p <- ggplot(data) + 
        # lat/long and group
        aes_string(x=longitude, y=latitude, group="group") + 
        # coloring based on the variable
        aes_string(fill=variable) +  geom_polygon() + 
        geom_path(color=path.color) + coord_equal() + 
        # mapping options
        mapOpts + 
        # x/y labels
        labs(x=xlab, y=ylab) + 
        # make the legend be titleless and have a long, tal color bar
        guides(fill=guide_colorbar(title=NULL, ticks=FALSE, barheight=barheight)) + 
        # make the color scale a gradient and use the chosen formatter
        scale_fill_gradient2(labels=formatter, space=space, high=fill.color.high) + 
        # title of plot
        opts(title=title, plot.title=theme_text(size=title.size, hjust=title.hjust)) +
        # faceting if called for
        facet
    
    return(p)
}


#' facet helper functions
#' 
#' Call facet functions
#' 
#' These functions merely serve to call \code{\link{facet_wrap}}, \code{\link{facet_grid}} or \code{\link{none}}.  It is needed because the program doesn't know if \code{link{facet_wrap}} or \code{\link{facet_grid}} is being called and hence doesn't know what arguments to pass, and those functions are not equiped with \dots.
#' 
#' @author Jared P. Lander
#' @aliases facet_wrap_helper facet_grid_helper none_helper
#' @seealso facet_wrap facet_grid
#' @import ggplot2
#' @param formula \code{\link{formula}} for use in \code{link{facet_wrap}}
#' @param nrow Number of rows.
#' @param ncol Number of columns.
#' @param scales Should scales be fixed ("fixed", the default), free ("free"), or free in one dimension ("free_x", "free_y").
#' @param \dots Catch all.
#' @return The result of \code{link{facet_wrap}}.
#' 
facet_wrap_helper <- function(formula, nrow=NULL, ncol=NULL, 
                              scales=c("fixed", "free", "free_y", "free_x"), ...)
{
    scales <- match.arg(scales)
    
    facet_wrap(formula, nrow=nrow, ncol=ncol, scales=scales)
}



facet_grid_helper <- function(formula, 
                              scales=c("fixed", "free", "free_y", "free_x"), ...)
{
    scales <- match.arg(scales)
    
    facet_grid(formula, scales=scales)
}


none_helper <- function(...)
{
    none(...)
}



#' none
#' 
#' Just Null
#' 
#' Returns NULL regardless of what is put in.
#' 
#' @author Jared P. Lander
#' @aliases none
#' @export none
#' @param \dots Anything.
#' @return \code{\link{NULL}} regardless of inputs.
#' @examples
#' 
#' none()
#' none(1)
#' none(mean)
#' 
none <- function(...)
{
    return(NULL)
}

#' map.options
#' 
#' Common mapping options
#' 
#' This builds a ggplot opts list that is commonly used for plotting.  Currently it only gives the default, perhaps I'll make it so users can pick and choose.
#' 
#' @author Jared P. Lander
#' @aliases map.options
#' @export map.options
#' @import ggplot2
#' @return A \code{\link{list}} of ggplot items.
map.options <- function()
{
    list(opts(panel.grid.major=theme_blank(), panel.grid.minor=theme_blank(), axis.text.x=theme_blank(), axis.text.y=theme_blank(), axis.ticks=theme_blank(), panel.background=theme_blank()))
}